libsnark/gadgetlib1/gadgets/fields/fp2_gadgets.tcc

```diff
@@ -107,6 +107,15 @@ Fp2_variable<Fp2T> Fp2_variable<Fp2T>::operator*(const FieldT &coeff) const
     return Fp2_variable<Fp2T>(this->pb, new_c0, new_c1, FMT(this->annotation_prefix, " operator*"));
 }
 
+template<typename Fp2T>
+Fp2_variable<Fp2T> Fp2_variable<Fp2T>::operator*(const Fp2T &fp2_const) const
```

Comment:
Why do you need this function? Shouldn't you use the `Fp2_mul_gadget` and do arithmetic with `Fp2_variable`s only? Where  do you invoke this version of the overloaded operator?

Reply:
Multiplication of an fp2_variable by a constant in Fp2 can be done as linear combinations without any mul gates (similar to mul by Fp constants or addition of 2 Fp2_variable), so this is a distinct operation from the mul gadget.  In terms of exactly where it's used, I use it in many places but the first one I can find is for multiplying by `b'^{-1}` (inverse of the b coeff for the twisted curve) in the `bls12_377_ate_dbl_gadget`.

Reply:
Well yes, here you do the operation without declaring new constraints in the system. My point was mostly why you didn't always created `Fp2_variable`s and used gadgets to do the arithmetic so that everything is constrained in the circuit.
I'll have a look at the gadget you refer to in Zecale to have more context, that's not a blocker to merge this PR.

Reply:
Unless I'm missing something, there should be no need to create any constraints.  The constant here just becomes the coefficients of a linear combination of variables, and these will appear at one of the inputs (or output) of one or more gates (so they are "burned into" the circuit).  In the very special case where we might need extra constraints, `generate_r1cs_equals_const_constraints` would probably be used in combination with this operator, but I'd guess that in most cases this is likely to be a waste of gates.

In the gadgets I'm creating, the values from fp2_const end up as coefficients of linear combinations used in other gates, and so everything is constrained by the circuit definition.

Reply:
I see. Great I'll have another look at the gadget you wrote in zecale